sdk/src/transaction_builder.rs

```diff
@@ -0,0 +1,310 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    move_types::{account_address::AccountAddress, language_storage::TypeTag},
+    types::{
+        account_config::{xdx_type_tag, xus_tag, XDX_NAME, XUS_NAME},
+        chain_id::ChainId,
+        transaction::{authenticator::AuthenticationKey, RawTransaction, TransactionPayload},
+    },
+};
+use std::fmt;
+
+pub use diem_transaction_builder::stdlib;
+
+pub struct TransactionBuilder {
+    sender: Option<AccountAddress>,
+    sequence_number: Option<u64>,
+    payload: TransactionPayload,
+    max_gas_amount: u64,
+    gas_unit_price: u64,
+    gas_currency_code: String,
+    expiration_timestamp_secs: u64,
+    chain_id: ChainId,
+}
+
+impl TransactionBuilder {
+    pub fn sender(mut self, sender: AccountAddress) -> Self {
+        self.sender = Some(sender);
+        self
+    }
+
+    pub fn sequence_number(mut self, sequence_number: u64) -> Self {
+        self.sequence_number = Some(sequence_number);
+        self
+    }
+
+    pub fn max_gas_amount(mut self, max_gas_amount: u64) -> Self {
+        self.max_gas_amount = max_gas_amount;
+        self
+    }
+
+    pub fn gas_unit_price(mut self, gas_unit_price: u64) -> Self {
+        self.gas_unit_price = gas_unit_price;
+        self
+    }
+
+    pub fn gas_currency_code<T: Into<String>>(mut self, gas_currency_code: T) -> Self {
+        self.gas_currency_code = gas_currency_code.into();
+        self
+    }
+
+    pub fn chain_id(mut self, chain_id: ChainId) -> Self {
+        self.chain_id = chain_id;
+        self
+    }
+
+    pub fn expiration_timestamp_secs(mut self, expiration_timestamp_secs: u64) -> Self {
+        self.expiration_timestamp_secs = expiration_timestamp_secs;
+        self
+    }
+
+    pub fn build(self) -> RawTransaction {
+        RawTransaction::new(
+            self.sender.expect("sender must have been set"),
+            self.sequence_number
+                .expect("sequence number must have been set"),
```

Comment:
if we expect to use this transaction builder in production code, we may consider never panic for validation

Reply:
another option is require sender and sequence_number in `build` arguments, if we don't want to return error

Reply:
Its ok to have code that can panic in production code, there's lots of stdlib functions that panic. If you forget to set the sender and you return an error, there's nothing that you can do to recover from that error since its a programmer error, it would require changing the code.